gadgets/mri_core/AutoScaleGadget.cpp

```diff
@@ -1,80 +1,64 @@
-int AutoScaleGadget::process_config(ACE_Message_Block* mb) {
-        max_value_ = max_value.value();
-	return GADGET_OK;
-}
+namespace {
+    template<class T>
+    Image<T> autoscale(Image<T> &image, float max_value, unsigned int histogram_bins) {
 
+		auto &header = std::get<ISMRMRD::ImageHeader>(image);
+        auto &data = std::get<hoNDArray<T>>(image);
+		auto &meta = std::get<optional<ISMRMRD::MetaContainer>>(image);
+		
+		if (header.image_type == ISMRMRD::ISMRMRD_IMTYPE_MAGNITUDE) { //Only scale magnitude images for now
+			auto max = *std::max_element(data.begin(), data.end());
+			auto current_scale_ = 1.0;
+			auto histogram_ = std::vector<size_t>(histogram_bins);
 
-int AutoScaleGadget::process(GadgetContainerMessage<ISMRMRD::ImageHeader> *m1, GadgetContainerMessage<hoNDArray<float> > *m2)
-{
-	if (m1->getObjectPtr()->image_type == ISMRMRD::ISMRMRD_IMTYPE_MAGNITUDE) { //Only scale magnitude images for now
-		float max = 0.0f;
-		float* d = m2->getObjectPtr()->get_data_ptr();
-		for (unsigned long int i = 0; i < m2->getObjectPtr()->get_number_of_elements(); i++) {
-			if (d[i] > max) max = d[i];
-		}
-
-		if (histogram_.size() != histogram_bins_) {
-			histogram_ = std::vector<size_t>(histogram_bins_);
-		}
-
-		for (size_t i = 0; i < histogram_bins_; i++) {
-			histogram_[i] = 0;
-		}
-
-		for (unsigned long int i = 0; i < m2->getObjectPtr()->get_number_of_elements(); i++) {
-			size_t bin = static_cast<size_t>(std::floor((d[i]/max)*histogram_bins_));
-			if (bin >= histogram_bins_) {
-				bin = histogram_bins_-1;
+			if (histogram_.size() != histogram_bins) {
+				histogram_ = std::vector<size_t>(histogram_bins);
 			}
-			histogram_[bin]++;
-		}
 
-		//Find 99th percentile
-		long long cumsum = 0;
-		size_t counter = 0;
-		while (cumsum < (0.99*m2->getObjectPtr()->get_number_of_elements())) {
-			cumsum += (long long)(histogram_[counter++]);
-		}
-		max = (counter+1)*(max/histogram_bins_);
-		GDEBUG("Max: %f\n",max);
+			for (size_t i = 0; i < histogram_bins; i++) {
+				histogram_[i] = 0;
+			}
 
-		current_scale_ = max_value_/max;
+			for (auto& d : data) {
+				size_t bin = static_cast<size_t>(std::floor((d/max)*histogram_bins));
+				if (bin >= histogram_bins) {
+					bin = histogram_bins-1;
+				}
+				histogram_[bin]++;
+			}
+			
+			//Find 99th percentile
+			long long cumsum = 0;
+			size_t counter = 0;
+			while (cumsum < (0.99*data.get_number_of_elements())) {
+				cumsum += (long long)(histogram_[counter++]);
+			}
+			max = (counter+1)*(max/histogram_bins);
+			GDEBUG("Max: %f\n",max);
+			current_scale_ = max_value/max;
 
-		for (unsigned long int i = 0; i < m2->getObjectPtr()->get_number_of_elements(); i++) {
-			d[i] *= current_scale_;
+			for (auto& d : data){
+				d *= current_scale_;
+			}
 		}
-	}
-
-	if (this->next()->putq(m1) < 0) {
-		GDEBUG("Failed to pass on data to next Gadget\n");
-		return GADGET_FAIL;
-	}
-
-	return GADGET_OK;
+        return Image<T>(header,data,meta);
+    }
+
+    template<class T>
```

Comment:
Can you do some template magic to make that substitution actual compile error.

Reply:
The way I was thinking about things, this error would be handling the case where someone fed in a complex image to the autoscale gadget because they messed up their pipeline xml, not knowing this gadget cant autoscale complex image types. The old code, if I understand it correctly, just "did nothing" for complex image types, which I feel like could be misleading if someone isn't looking at the code and thinks the images are now scaled. I wanted to bubble an error that the pipeline isn't doing what they think it is, but I don't think this can be a compile time error because the image type isn't going to be known until the pipeline is defined? 

The other option I'd see is to only accept non-complex image inputs, but that doesn't match the old behavior, and similarly might break existing pipelines that are "autoscaling (with no effect)" complex images.